Fix `this` being captured by immediately invoked functions

chengluyu opened the pull request:
This is mentioned in #65 .

```
:js
class A: { x: int }
  method MA = let _ = 1 in this.x
//│ Defined class A
//│ Defined A.MA: A -> int
//│ // Prelude
//│ class A {
//│   constructor(fields) {
//│     this.x = fields.x;
//│   }
//│   get MA() {
//│     return ((function (_) {
//│       return this.x;
//│     })(1));
//│   }
//│ }
//│ // End of generated code

(A { x = 1 }).MA
//│ res: int
//│ Runtime error:
//│   TypeError: Cannot read properties of undefined (reading 'x')
```

chengluyu committed:
Fix `this` being captured by immediately invoked functions

LPTK commented:
Cool!

But does this really solve the problem in general? What if there is a captured `this` in a generated local `function`?

chengluyu commented:
Sadly this PR only works in this particular case.

The root of the problem is we implement `let` and pattern matches with IIFEs. IIFEs introduce extra `function` scopes in generated code, which shadow `this`. Although we can keep track of runtime scopes and add alias variable declarations but that would be cumbersome.

A better solution is to avoid IIFEs and generate a series of JavaScript statements.

LPTK commented:
Why not just add a `const self = this` at the beginning of methods and then refer to this `self` instead? It should be a trivial change that would avoid mistranslating the semantics of MLscript.

LPTK commented:
Of course we need to make sure that `self` itself is not shadowed. I think your `Scope` infrastructure should be able to deal with that.

Also, if the `self` symbol is never accessed in the method body, we can omit the binding in the generated code.

chengluyu committed:
Add `const self = this` in the beginning of class members

chengluyu commented:
It's done.

LPTK commented:
Please ad some tests making sure that this `self` cannot be shadowed.

chengluyu committed:
Fix the case where `this` is shadowed in class methods

chengluyu commented:
I added some test cases but found that shadowing `this` caused problems. The reason is that we didn't allocate runtime names for parameters (as well as parameter patterns) in `translatePattern`. The latest commit fixes this. I also added test cases for illegal parameters.

LPTK commented:
Ok, now that https://github.com/hkust-taco/mlscript/pull/117 is merged you can adapt your changes here and we can merge it.

chengluyu committed:
Fix that `declareParameter` does not allocate runtime names

... plus a test case in `Classes` causes type errors

chengluyu commented:
I rebased the branch. This PR is ready.

LPTK commented on shared/src/main/scala/mlscript/codegen/Scope.scala:
```diff
@@ -51,6 +51,16 @@ class Scope(name: Str, enclosing: Opt[Scope]) {
+  /**
+    * Shorthands for creating function scopes.
+    */
+  def this(name: Str, params: Ls[Str], enclosing: Scope) = {
```
Is this used?

LPTK commented on shared/src/test/diff/mlscript/TypeClasses.mls:
```diff
@@ -1,4 +1,3 @@
-
```
Let's not add useless space changes.

LPTK replied:
(I actually dislike when tests start immediately on the first line. We separate blocks by empty lines, so this feels wrong.)

LPTK reviewed (approved):
Ok, LGTM.

But it actually occurred to me that we could also define `self` in the class scope, instead of in each method, which would produce leaner code. As in:

```js
class C { self = this; ... }
```

Would that be easy to change now?

chengluyu committed:
Fix minor issues mentioned in the PR

chengluyu commented:
> Ok, LGTM.
> 
> But it actually occurred to me that we could also define `self` in the class scope, instead of in each method, which would produce leaner code. As in:
> 
> ```js
> class C { self = this; ... }
> ```
> 
> Would that be easy to change now?

There had been some trouble implementing this. Now I think `self = this` is a temporary and ugly fix. It will be unnecessary after we remove IIFEs. So let’s not do it right now.

chengluyu committed:
Add line number changes in TypeClasses.mls

LPTK commented on shared/src/main/scala/mlscript/codegen/Scope.scala:
```diff
@@ -51,16 +51,6 @@ class Scope(name: Str, enclosing: Opt[Scope]) {
-      declareValue(param)
-    }
-  }
-
```
See, this is exactly what I was telling you about being careful with merges... It is not enough to just click "accept both". Please be more mindful next time.

chengluyu replied:
👀 Sorry, I didn't get. I deleted this because of [this comment](https://github.com/hkust-taco/mlscript/pull/123#discussion_r927398129). I didn't click any “accept both” button we deleting this.

LPTK replied:
The point is that this constructor was deleted on MLScript's main branch. It was deleted in https://github.com/hkust-taco/mlscript/pull/117, on which you rebased your changes. You reintroduced it because you botched your merge. Please be more careful.

LPTK commented:
> Now I think self = this is a temporary and ugly fix. It will be unnecessary after we remove IIFEs.

I disagree. JS semantics is just stupid and we should NOT follow it, so the self-fix will be needed in general.

The following MLscript (using the future syntax):

```fsharp
class A {
  fun m() =
    fun f() = this.x
    f
}
```
should generate this JS:

```js
class A {
  m() {
    let self = this
    function f() { return self.x }
    return f
  }
}
```

or alternatively:

```js
class A {
  #self = this
  m() {
    function f() { return #self.x }
    return f
  }
}
```

But it's okay if we don't do the latter for now if it's too annoying to implement.

LPTK commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -388,20 +398,43 @@ class JSBackend {
+      case _ => scope.derive(s"getter $name")
+    }
+    // Declare the alias for `this` before declaring parameters.
+    // TODO: Remove this after we don't use IIFEs.
```
```suggestion
```

chengluyu commented:
> > Now I think self = this is a temporary and ugly fix. It will be unnecessary after we remove IIFEs.
> 
> I disagree. JS semantics is just stupid and we should NOT follow it, so the self-fix will be needed in general.
> 
> The following MLscript (using the future syntax):
> 
> ```fsharp
> class A {
>   fun m() =
>     fun f() = this.x
>     f
> }
> ```
> 
> should generate this JS:
> 
> ```js
> class A {
>   m() {
>     let self = this
>     function f() { return self.x }
>     return f
>   }
> }
> ```
> 
> or alternatively:
> 
> ```js
> class A {
>   #self = this
>   m() {
>     function f() { return #self.x }
>     return f
>   }
> }
> ```
> 
> But it's okay if we don't do the latter for now if it's too annoying to implement.

This is another case. It’s not IIFE. In your case, we should create alias because of the nested function `f`. It should only be created when the body of `f` refers to `this`.

But now we add the alias anyway. That's why I say it’s ugly because the generated code is not easy to understand. In the future, we will deprecate IIFEs for pattern matching and `let` expressions. Then we can remove the fix in this PR. By the time when the nested function syntax arrives, we should add anther mechanism which prevents inner functions shadowing `this`.

chengluyu committed:
Update shared/src/main/scala/mlscript/JSBackend.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

LPTK commented:
> By the time when the nested function syntax arrives, we should add anther mechanism which prevents inner functions shadowing this.

But why add another mechanism? It's the same problem. The fix could be _improved_ later, but there is no reason to just remove it.
